File: sdks/python/starfish/session.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .connection import Connection
from .emitter import Observable
from .id import next_id
from .types import StarfishFrame, StarfishHeader


@dataclass
class JoinOptions:
    name: str | None = None
    role: str | None = None
    meta: dict[str, Any] | None = None
    create: bool = True


@dataclass
class ClientInfo:
    id: str
    name: str | None = None
    role: str | None = None
    meta: dict[str, Any] = field(default_factory=dict)


class Session:
    def __init__(self, connection: Connection) -> None:
        self._connection = connection
        self._session: str | None = None
        self._clients: dict[str, ClientInfo] = {}

        self.clients: Observable[list[ClientInfo]] = Observable([])
        self.peers: Observable[list[ClientInfo]] = Observable([])

    @property
    def current(self) -> str | None:
        return self._session

    @property
    def client_id(self) -> str | None:
        return self._connection.client_id

    async def join(self, session: str, options: JoinOptions | None = None) -> StarfishFrame:
        opts = options or JoinOptions()
        frame = StarfishFrame(
            header=StarfishHeader(
                id=next_id("join"),
                resource="session",
                method="join",
                kind="request",
                session=session,
            ),
            payload={
                "create": opts.create,
                "name": opts.name or self._connection.client_id or "client",
                "role": opts.role or "default",
                "meta": opts.meta or {},
            },
        )

        response = await self._connection.send_and_wait(frame)
        self._session = session

        clients: list[dict[str, Any]] = (response.payload or {}).get("clients", [])
        self._clients.clear()
        for c in clients:
            info = ClientInfo(
                id=c["id"],
                name=c.get("name"),
                role=c.get("role"),
                meta=c.get("meta", {}),
            )
            self._clients[info.id] = info
        self._update_observables()

        return response

    async def leave(self) -> None:
        if not self._session:
            return

        frame = StarfishFrame(
            header=StarfishHeader(
                id=next_id("leave"),
                resource="session",
                method="leave",
                kind="request",
                session=self._session,
            ),
        )

        await self._connection.send(frame)
        self._session = None
        self._clients.clear()
        self._update_observables()

    def handle_frame(self, frame: StarfishFrame) -> None:
        if not self._session or frame.header.session != self._session:
            return

        if frame.header.resource != "session" or frame.header.kind != "event":
            return

        if frame.header.method == "connected":
            client_data = (frame.payload or {}).get("client")
            if client_data:
                info = ClientInfo(
                    id=client_data["id"],
                    name=client_data.get("name"),
                    role=client_data.get("role"),
                    meta=client_data.get("meta", {}),
                )
                self._clients[info.id] = info
                self._update_observables()

        elif frame.header.method == "disconnected":
            client_id = (frame.payload or {}).get("clientId")
            if client_id and client_id in self._clients:
                del self._clients[client_id]
                self._update_observables()

    def _update_observables(self) -> None:
        all_clients = list(self._clients.values())
        self.clients.set(all_clients)
        self.peers.set([c for c in all_clients if c.id != self.client_id])

File: sdks/golang/starfish/session.go
package starfish

import (
	"context"
	"sync"
	"time"
)

// JoinOptions configures a session join request.
type JoinOptions struct {
	Create bool
	Name   string
	Role   string
	Meta   map[string]any
}

// JoinResult is the server's response to a session join.
type JoinResult struct {
	ClientID string
	Clients  []ClientInfo
}

// sessionManager handles session join/leave and client tracking.
type sessionManager struct {
	mu      sync.RWMutex
	conn    *connection
	idg     *IDGenerator
	session string
	clients []ClientInfo
}

func newSessionManager(conn *connection, idg *IDGenerator) *sessionManager {
	return &sessionManager{
		conn: conn,
		idg:  idg,
	}
}

// join sends a session.join request and returns the result.
func (s *sessionManager) join(ctx context.Context, session string, opts *JoinOptions) (*JoinResult, error) {
	ts := time.Now().UnixMilli()
	payload := map[string]any{}
	if opts != nil {
		if opts.Create {
			payload["create"] = true
		}
		if opts.Name != "" {
			payload["name"] = opts.Name
		}
		if opts.Role != "" {
			payload["role"] = opts.Role
		}
		if opts.Meta != nil {
			payload["meta"] = opts.Meta
		}
	}

	frame := &Frame{
		Header: Header{
			V:        2,
			ID:       s.idg.Next("session"),
			Resource: "session",
			Method:   "join",
			Kind:     "request",
			Session:  session,
			Ts:       &ts,
		},
		Payload: payload,
	}

	reply, err := s.conn.sendAndWait(ctx, frame, 0)
	if err != nil {
		return nil, err
	}

	result := &JoinResult{}
	if v, ok := reply.Payload["clientId"].(string); ok {
		result.ClientID = v
	}
	if clientsRaw, ok := reply.Payload["clients"].([]any); ok {
		for _, c := range clientsRaw {
			if cm, ok := c.(map[string]any); ok {
				info := ClientInfo{}
				if v, ok := cm["id"].(string); ok {
					info.ID = v
				}
				if v, ok := cm["name"].(string); ok {
					info.Name = v
				}
				if v, ok := cm["role"].(string); ok {
					info.Role = v
				}
				if v, ok := cm["meta"].(map[string]any); ok {
					info.Meta = v
				}
				result.Clients = append(result.Clients, info)
			}
		}
	}

	s.mu.Lock()
	s.session = session
	s.clients = result.Clients
	s.mu.Unlock()

	return result, nil
}

// leave sends a session.leave request.
func (s *sessionManager) leave(ctx context.Context) error {
	s.mu.RLock()
	session := s.session
	s.mu.RUnlock()

	if session == "" {
		return nil
	}

	ts := time.Now().UnixMilli()
	frame := &Frame{
		Header: Header{
			V:        2,
			ID:       s.idg.Next("session"),
			Resource: "session",
			Method:   "leave",
			Kind:     "request",
			Session:  session,
			Ts:       &ts,
		},
	}

	err := s.conn.send(ctx, frame)

	s.mu.Lock()
	s.session = ""
	s.clients = nil
	s.mu.Unlock()

	return err
}

// handleFrame processes session-related events.
func (s *sessionManager) handleFrame(f *Frame) {
	if f.Header.Resource != "client" {
		return
	}

	switch f.Header.Method {
	case "connected":
		if clientRaw, ok := f.Payload["client"].(map[string]any); ok {
			info := ClientInfo{}
			if v, ok := clientRaw["id"].(string); ok {
				info.ID = v
			}
			if v, ok := clientRaw["name"].(string); ok {
				info.Name = v
			}
			if v, ok := clientRaw["role"].(string); ok {
				info.Role = v
			}
			if v, ok := clientRaw["meta"].(map[string]any); ok {
				info.Meta = v
			}
			s.mu.Lock()
			s.clients = append(s.clients, info)
			s.mu.Unlock()
		}
	case "disconnected":
		if clientID, ok := f.Payload["clientId"].(string); ok {
			s.mu.Lock()
			for i, c := range s.clients {
				if c.ID == clientID {
					s.clients = append(s.clients[:i], s.clients[i+1:]...)
					break
				}
			}
			s.mu.Unlock()
		}
	}
}

// currentSession returns the current session name.
func (s *sessionManager) currentSession() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.session
}

// getClients returns the current client list.
func (s *sessionManager) getClients() []ClientInfo {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]ClientInfo, len(s.clients))
	copy(out, s.clients)
	return out
}
